File: fetch_yahoo.py
#!/usr/bin/env python3
"""
fetch_yahoo.py  — fetch closing prices from Yahoo Finance chart API.
Called by server.js:  python3 fetch_yahoo.py SYMBOL [RANGE]
Prints JSON to stdout: {"ok":true,"symbol":"AAPL","data":[["2026-07-10",100.0],...]}

Uses Python's urllib which has a different TLS fingerprint than Node.js,
bypassing Yahoo's bot-detection that blocks Node.js.
"""
import sys, json, urllib.request, urllib.error, ssl, time, random

def fetch(symbol, range_='5d'):
    ctx = ssl.create_default_context()
    encoded = urllib.request.quote(symbol, safe='')
    host = random.choice(['query1', 'query2'])
    url = f'https://{host}.finance.yahoo.com/v8/finance/chart/{encoded}?range={range_}&interval=1d&includePrePost=false'
    headers = {
        'User-Agent': 'Mozilla/5.0',
        'Accept': '*/*',
        'Referer': 'https://finance.yahoo.com',
    }
    req = urllib.request.Request(url, headers=headers)
    try:
        with urllib.request.urlopen(req, context=ctx, timeout=15) as resp:
            raw = json.loads(resp.read())
    except urllib.error.HTTPError as e:
        return {'ok': False, 'error': f'HTTP {e.code}'}
    except Exception as e:
        return {'ok': False, 'error': str(e)}

    try:
        result = raw['chart']['result'][0]
        timestamps = result['timestamp']
        closes = result['indicators']['quote'][0]['close']
        data = []
        for ts, cl in zip(timestamps, closes):
            if cl is not None and cl == cl:  # not None, not NaN
                date = time.strftime('%Y-%m-%d', time.gmtime(ts))
                data.append([date, round(cl, 4)])
        return {'ok': True, 'symbol': symbol, 'data': data}
    except Exception as e:
        return {'ok': False, 'error': f'parse: {e}'}

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print(json.dumps({'ok': False, 'error': 'usage: fetch_yahoo.py SYMBOL [RANGE]'}))
        sys.exit(1)
    symbol = sys.argv[1]
    range_ = sys.argv[2] if len(sys.argv) > 2 else '5d'
    print(json.dumps(fetch(symbol, range_)))

File: fetch_yahoo.js
const { chromium } = require('playwright');
const symbol = process.argv[2];
const range = process.argv[3] || '5d';

if (!symbol) {
  console.log(JSON.stringify({ ok: false, error: 'usage: node fetch_yahoo.js SYMBOL [RANGE]' }));
  process.exit(1);
}

// CTA ETF proxies must use adjusted closes so dividends/splits do not create
// artificial trend breaks. Existing index/futures/dashboard symbols keep the
// historical raw-close behavior of this helper.
const ADJUSTED_CLOSE_SYMBOLS = new Set(['SPY', 'QQQ', 'IWM', 'IEF', 'USO', 'GLD', 'TLT', 'SOXX']);

(async () => {
  let browser;
  try {
    browser = await chromium.launch({ headless: true });
    const context = await browser.newContext({
      userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
    });
    const page = await context.newPage();

    // Go to Yahoo Finance main page to set cookies/crumb
    await page.goto(`https://finance.yahoo.com/quote/${symbol}`, { waitUntil: 'domcontentloaded', timeout: 15000 }).catch(() => {});

    // Now fetch the JSON API. Include corporate actions / adjusted close so ETF
    // trend proxies can remain on a total-return-consistent price history.
    const url = `https://query1.finance.yahoo.com/v8/finance/chart/${encodeURIComponent(symbol)}?range=${range}&interval=1d&includePrePost=false&events=div%2Csplits&includeAdjustedClose=true`;
    const response = await page.goto(url, { waitUntil: 'domcontentloaded', timeout: 15000 });
    
    if (!response || !response.ok()) {
      console.log(JSON.stringify({ ok: false, error: `HTTP ${response ? response.status() : 'Unknown'}` }));
      process.exit(0);
    }
    
    const text = await response.text();
    const raw = JSON.parse(text);
    
    const result = raw.chart.result[0];
    const timestamps = result.timestamp;
    const closes = result.indicators.quote[0].close;
    const adjCloses = result.indicators?.adjclose?.[0]?.adjclose || closes;
    const useAdjusted = ADJUSTED_CLOSE_SYMBOLS.has(symbol);
    
    const data = [];
    if (timestamps && closes) {
      for (let i = 0; i < timestamps.length; i++) {
        const ts = timestamps[i];
        const rawClose = closes[i];
        const adjustedClose = adjCloses ? adjCloses[i] : null;
        const selected = useAdjusted && adjustedClose != null ? adjustedClose : rawClose;
        if (selected != null && !Number.isNaN(selected)) {
          const dateStr = new Date(ts * 1000).toISOString().split('T')[0];
          data.push([dateStr, Number(selected.toFixed(4))]);
        }
      }
    }
    
    console.log(JSON.stringify({ ok: true, symbol, priceField: useAdjusted ? 'adjClose' : 'close', data }));
  } catch (err) {
    console.log(JSON.stringify({ ok: false, error: err.message }));
  } finally {
    if (browser) await browser.close();
  }
})();
